src/mlpack/core/optimizers/bigbatch_sgd/bigbatch_sgd_impl.hpp

```diff
@@ -0,0 +1,250 @@
+double BigBatchSGD<UpdatePolicyType>::Optimize(
+    DecomposableFunctionType& function, arma::mat& iterate)
+{
+  // Find the number of functions.
+  const size_t numFunctions = function.NumFunctions();
+  size_t numBatches = numFunctions / batchSize;
+  if (numFunctions % batchSize != 0)
+    ++numBatches; // Capture last few.
+
+  // Batch visitation order.
+  arma::Col<size_t> visitationOrder = arma::linspace<arma::Col<size_t>>(0,
+      (numBatches - 1), numBatches);
+
+  if (shuffle)
+    visitationOrder = arma::shuffle(visitationOrder);
+
+  // To keep track of where we are and how things are going.
+  std::vector<arma::mat> funcGradients(batchSize);
+  size_t currentBatch = 0;
+  double overallObjective = 0;
+  double lastObjective = DBL_MAX;
+  double vB = 0;
+  double gB = 0;
+  bool reset = false;
+
+  // Calculate the first objective function.
+  for (size_t i = 0; i < numFunctions; ++i)
+    overallObjective += function.Evaluate(iterate, i);
+
+  // Now iterate!
+  arma::mat gradient(iterate.n_rows, iterate.n_cols);
+  for (size_t i = 1; i != maxIterations; ++i, ++currentBatch)
+  {
+    // Is this iteration the start of a sequence?
+    if ((currentBatch % numBatches) == 0)
+    {
+      // Output current objective function.
+      Log::Info << "Big-batch SGD: iteration " << i << ", objective "
+          << overallObjective << "." << std::endl;
+
+      if (std::isnan(overallObjective) || std::isinf(overallObjective))
+      {
+        Log::Warn << "Big-batch SGD: converged to " << overallObjective
+            << "; terminating with failure.  Try a smaller step size?"
+            << std::endl;
+        return overallObjective;
+      }
+
+      if (std::abs(lastObjective - overallObjective) < tolerance)
+      {
+        Log::Info << "Big-batch SGD: minimized within tolerance " << tolerance
+            << "; terminating optimization." << std::endl;
+        return overallObjective;
+      }
+
+      // Reset the counter variables.
+      lastObjective = overallObjective;
+      overallObjective = 0;
+      currentBatch = 0;
+
+      if (shuffle)
+        visitationOrder = arma::shuffle(visitationOrder);
+    }
+
+    const size_t offset = batchSize * visitationOrder[currentBatch];
+    if (visitationOrder[currentBatch] != numBatches - 1)
+    {
+      // Compute the stochastic gradient estimation.
+      function.Gradient(iterate, offset, funcGradients[0]);
+      gradient = funcGradients[0];
+      for (size_t j = 1; j < batchSize; ++j)
+      {
+        function.Gradient(iterate, offset + j, funcGradients[j]);
+        gradient += funcGradients[j];
+      }
+      gB = std::pow(arma::norm(gradient / batchSize, 2), 2.0);
+
+      // Compute sample variance.
+      vB = std::pow(arma::norm(funcGradients[0] - (gradient / batchSize),
+          2.0), 2.0);
+      for (size_t j = 1; j < batchSize; ++j)
+      {
+        vB += std::pow(arma::norm(funcGradients[j] - (gradient / batchSize),
+            2.0), 2.0);
+      }
```

Comment:
I think it might be possible to avoid holding the full vector `funcGradients` by using an incremental variance calculation, something like this:

https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Online_algorithm

However I am not totally sure.  But for large batch sizes, I think the whole algorithm could be much faster if all the memory for `funcGradients` was not needed.  I think for large batch sizes and large networks being optimized, that might actually cause the system to run out of RAM in some cases.